pkg/validator/promql_expression.go

```diff
@@ -121,6 +121,47 @@ func (h expressionDoesNotUseLabels) Validate(_ unmarshaler.RuleGroup, rule rulef
 	return errs
 }
 
+type expressionUsesOnlyAllowedLabelsForMetricRegexp struct {
+	allowedLabels    []string
+	metricNameRegexp string
+}
+
+func newExpressionUseOnlyWhitelistedLabelsForMetric(paramsConfig yaml.Node) (Validator, error) {
+	params := struct {
+		AllowedLabels    []string `yaml:"allowedLabels"`
+		MetricNameRegexp string   `yaml:"metricNameRegexp"`
+	}{}
+	if err := paramsConfig.Decode(&params); err != nil {
+		return nil, err
+	}
+	if len(params.AllowedLabels) == 0 {
+		return nil, fmt.Errorf("missing labels")
+	}
+	return &expressionUsesOnlyAllowedLabelsForMetricRegexp{allowedLabels: params.AllowedLabels, metricNameRegexp: params.MetricNameRegexp}, nil
```

Comment:
I completely missed this first time, but please make the `expressionUsesOnlyAllowedLabelsForMetricRegexp.metricNameRegexp` a `regexp.Regexp` type and compile it in the `newExpressionUseOnlyWhitelistedLabelsForMetric` so invalid regexp is caught during unmarshalling? This wait it could result in panic because of user input.

Same as here for example https://github.com/FUSAKLA/promruval/blob/414a2fc6008248c17fdef794f0d0f3f218a0f1ff/pkg/validator/others.go#L20

Also make the regexp anchored already in the `newExpressionUseOnlyWhitelistedLabelsForMetric` function and I see the `regexp.Compile("^" + regexp + "$")` in multiple places ([here](https://github.com/FUSAKLA/promruval/blob/414a2fc6008248c17fdef794f0d0f3f218a0f1ff/pkg/validator/promql_expression.go#L456) [here](https://github.com/FUSAKLA/promruval/blob/414a2fc6008248c17fdef794f0d0f3f218a0f1ff/pkg/validator/others.go#L20)). Might be worth add a `compileAnchoredRegexp(string) (regexp.Regexp, error)` function?

Reply:
Refactored as proposed.